AndroidHomeApplication/app/src/main/java/com/example/androidhomeapplication/data/repository/MoviesRepository.kt

```diff
@@ -0,0 +1,112 @@
+package com.example.androidhomeapplication.data.repository
+
+import com.example.androidhomeapplication.data.models.Genre
+import com.example.androidhomeapplication.data.models.Movie
+import com.example.androidhomeapplication.data.models.MovieDetails
+import com.example.androidhomeapplication.data.remote.RetrofitBuilder
+import com.example.androidhomeapplication.data.remote.response.*
+import com.example.androidhomeapplication.data.remote.services.ConfigurationService
+import com.example.androidhomeapplication.data.remote.services.MoviesService
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import retrofit2.Retrofit
+
+private const val SECURE_BASE_URL: String = "https://image.tmdb.org/t/p/"
+
+enum class ImageType(val size: String) {
+    POSTER(size = "w500"),
+    BACKDROP(size = "w780"),
+    PROFILE(size = "w185")
+}
+
+class MoviesRepository() {
+    private val retrofit: Retrofit = RetrofitBuilder.buildRetrofit()
+    private val configurationService: ConfigurationService by lazy {
+        retrofit.create(
+            ConfigurationService::class.java
+        )
+    }
+    private val movieService: MoviesService by lazy { retrofit.create(MoviesService::class.java) }
+
+    private val mutex = Mutex()
+    private var genres: Map<Long, Genre>? = null
+    private var configInfo: ConfigurationResponse? = null
+
+    suspend fun loadMovieById(movieId: Long): MovieDetails = coroutineScope {
+        val profileImageUrl = getImageUrlByType(ImageType.PROFILE)
+        val backdropImageUrl = getImageUrlByType(ImageType.BACKDROP)
+
+        val details = async { movieService.loadMovieDetails(movieId) }
+        val casts = async {
+            movieService.loadMovieCredits(movieId).cast.map { castResponse ->
+                castResponse.mapToActor(profileImageUrl)
+            }
+        }
+
+        details.await().mapToMovieDetails(backdropImageUrl, casts.await())
+    }
+
+    suspend fun loadMovies(queryString: String?, page: Int): List<Movie> = coroutineScope {
+        val movies = async { loadData(queryString, page) }
+        val genresMap = async { getCachedGenresOrLoad() }
+        val posterImageUrl = getImageUrlByType(ImageType.POSTER)
+
+        movies.await().map { movie ->
+            val movieGenres =
+                movie.genreIds.mapNotNull { singleGenreId -> genresMap.await()[singleGenreId] }
```

Comment:
looks not nice when u call .await() for each .map { movie ->
save genres.await() result before calling map for movies